File: app/src/main/java/com/cazimir/relaxoo/ui/sound_grid/SoundGridFragment.java
package com.cazimir.relaxoo.ui.sound_grid;

import android.app.TimePickerDialog;
import android.content.Context;
import android.media.SoundPool;
import android.os.Bundle;
import android.os.CountDownTimer;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.GridView;
import android.widget.ImageButton;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;

import com.cazimir.relaxoo.R;
import com.cazimir.relaxoo.adapter.GridAdapter;
import com.cazimir.relaxoo.dialog.TimerDialog;
import com.cazimir.relaxoo.model.SavedCombo;
import com.cazimir.relaxoo.model.Sound;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.TimeUnit;

import butterknife.BindView;
import butterknife.ButterKnife;

public class SoundGridFragment extends Fragment {

  private static final String TAG = "SoundGridFragment";

  @BindView(R.id.upper_buttons)
  LinearLayout upperButtons;

  @BindView(R.id.gridView)
  GridView gridView;

  @BindView(R.id.mute_button)
  ImageButton muteButton;

  @BindView(R.id.random_button)
  ImageButton randomButton;

  @BindView(R.id.play_button)
  ImageButton playStopButton;

  @BindView(R.id.save_fav_button)
  ImageButton saveFavorites;

  @BindView(R.id.set_timer_button)
  ImageButton setTimer;

  @BindView(R.id.timerText)
  TextView timerText;

  private GridAdapter gridArrayAdapter;
  private SoundPool soundPool;
  private SoundGridViewModel viewModel;

  private boolean muted;
  private OnActivityCallback activityCallback;

  private MutableLiveData<Boolean> timerEnabled = new MutableLiveData<>();
  private CountDownTimer countDownTimer;

  private TimePickerDialog.OnTimeSetListener timerPickerListener =
          (view, hours, minutes) -> toggleCountdownTimer((hours * 60) + minutes);

  public static SoundGridFragment newInstance() {
    return new SoundGridFragment();
  }

  @Nullable
  @Override
  public View onCreateView(
      @NonNull LayoutInflater inflater,
      @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {

    View view = inflater.inflate(R.layout.sound_list_fragment, container, false);

    ButterKnife.bind(this, view);

    return view;
  }

  private void playStopSound(int soundPoolId, boolean playing, int streamId) {
    Log.d(TAG, "playStopSound: called");
    if (playing) {
      Log.d(TAG, "stopping sound");
      soundPool.stop(streamId);
      // update viewmodel for favorites fragment somehow - perhaps through activity?

      viewModel.updateSoundList(soundPoolId, 0);
    } else {
      int newStreamId = soundPool.play(soundPoolId, 0.5f, 0.5f, 0, -1, 1);
      Log.d(TAG, "playing sound");
      viewModel.updateSoundList(soundPoolId, newStreamId);
    }
  }

  @Override
  @NonNull
  public void onAttach(Context context) {
    super.onAttach(context);
    Log.d(TAG, "SoundGridFragment onAttach() called with: ");
    activityCallback = (OnActivityCallback) context;
  }

  private boolean stopAllSounds() {

    Log.d(TAG, "stopAllSounds: called");

    List<Sound> list = new CopyOnWriteArrayList<>(viewModel.playingSounds().getValue());

    // TODO: 22.12.2019 try to remove some of the for loops, use guava or something
    for (Sound sound : list) {
      soundPool.stop(sound.streamId());
      viewModel.updateSoundList(sound.soundPoolId(), -1);
    }

    return false;
  }

  @Override
  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
    super.onActivityCreated(savedInstanceState);
    viewModel = ViewModelProviders.of(this).get(SoundGridViewModel.class);

    soundPool = viewModel.createOrGetSoundPool();

    setListenersForButtons();

    soundPool.setOnLoadCompleteListener(
            (soundPool, sampleId, status) -> {
              Log.d(TAG, "onLoadComplete: " + sampleId);
              viewModel.addedSound();
            });

    // region Observers

    // change icon to unmute
    viewModel
        .mutedLiveData()
        .observe(
            getViewLifecycleOwner(),
            new Observer<Boolean>() {
              @Override
              public void onChanged(Boolean muted) {
                if (muted) {
                  muteButton.setImageDrawable(getResources().getDrawable(R.drawable.ic_mute_off));

                } else {
                  muteButton.setImageDrawable(getResources().getDrawable(R.drawable.ic_mute_on));
                }
              }
            });

    // listen if at least one sound is playing to show notification in actionbar to user
    viewModel
        .isAtLeastOneSoundPlaying()
        .observe(
            getViewLifecycleOwner(),
            new Observer<Boolean>() {
              @Override
              public void onChanged(Boolean playing) {

                if (playing) {
                  activityCallback.showNotification();
                } else {
                  activityCallback.hideNotification();
                }
              }
            });

    // listen for changes to the sound lists live data object to set the adapter for the gridview
    // along with the callback methods (clicked & volume changed)
    viewModel
        .sounds()
        .observe(
            // TODO: 19.12.2019 move in a separate file or inner class
            getViewLifecycleOwner(),
                sounds -> {
                  Log.d(TAG, "Sound list changed: " + sounds);

                  if (!sounds.isEmpty()) {
                    activityCallback.soundsFetchedAndSaved();
                  }
                  gridArrayAdapter =
                          new GridAdapter(
                                  getContext(),
                                  sounds,
                                  new OnSoundClickListener() {
                                    @Override
                                    public void clicked(
                                            int soundId, boolean playing, int streamId, boolean pro) {

                                      if (pro) {
                                        activityCallback.showBottomDialogForPro();
                                      } else {
                                        playStopSound(soundId, playing, streamId);
                                      }
                                    }

                                    @Override
                                    public void volumeChange(int streamId, int progress) {
                                      // TODO: 18.12.2019 refactor this float to string to double
                                      // transformation
                                      Log.d(
                                              TAG, "volumeChange: called with volume: " + (double) progress / 100);
                                      soundPool.setVolume(
                                              streamId,
                                              Float.valueOf(String.valueOf((double) progress / 100)),
                                              Float.valueOf(String.valueOf((double) progress / 100)));
                                    }
                                  });
                  //              }

                  gridView.setAdapter(gridArrayAdapter);

                  // if sound not loaded yet and sounds list not yet populated
                  if (!sounds.isEmpty() && atLeastOneSoundWithoutSoundPoolId(sounds)) {
                    loadToSoundPool(sounds);
                  }
                });

    // listen to the playing sounds live data object to change the play stop getLogoPath icon on top
    viewModel
        .playingSounds()
        .observe(
            getViewLifecycleOwner(),
            new Observer<List<Sound>>() {
              @Override
              public void onChanged(List<Sound> playingList) {
                if (playingList.isEmpty()) {
                  playStopButton.setImageDrawable(getResources().getDrawable(R.drawable.ic_play));
                  setTimer.setImageDrawable(getResources().getDrawable(R.drawable.ic_timer_on));
                  if (countDownTimer != null) {
                    timerEnabled.setValue(false);
                  }
                } else {
                  playStopButton.setImageDrawable(getResources().getDrawable(R.drawable.ic_stop));
                }
              }
            });

    // TODO: 06-Jan-20 move this to viewmodel
    timerEnabled.observe(
        getViewLifecycleOwner(),
        new Observer<Boolean>() {
          @Override
          public void onChanged(Boolean isTimerEnabled) {

            Log.d(TAG, "timerEnabled: " + isTimerEnabled);

            if (isTimerEnabled) {
              showTimerText();
              setTimer.setImageDrawable(getResources().getDrawable(R.drawable.ic_timer_off));
            } else {
              hideTimerText();
              setTimer.setImageDrawable(getResources().getDrawable(R.drawable.ic_timer_on));
              countDownTimer.cancel();
              Log.d(TAG, "countDownTimer canceled!");
            }
          }
        });

    viewModel
            .getSoundsLoadedToSoundPool()
            .observe(
                    getViewLifecycleOwner(),
                    new Observer<Integer>() {
                      @Override
                      public void onChanged(Integer soundsAdded) {

                        if (viewModel.sounds().getValue().size() != 0) {
                  if (soundsAdded == viewModel.sounds().getValue().size()) {
                    activityCallback.hideSplash();
                  }
                        }
                      }
                    });

    // endregion
  }

  private boolean atLeastOneSoundWithoutSoundPoolId(List<Sound> sounds) {
    for (Sound sound : sounds) {
      if (sound.soundPoolId() == 0) {
        return true;
      }
    }

    return false;
  }

  // region Listeners for buttons on top
  private void setListenersForButtons() {

    Log.d(TAG, "setListenersForButtons: called");

    muteButton.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View v) {

            muted = !muted;

            if (muted) {
              for (Sound sound : viewModel.playingSounds().getValue()) {
                soundPool.setVolume(sound.streamId(), 0f, 0f);
              }
            } else {
              for (Sound sound : viewModel.playingSounds().getValue()) {
                soundPool.setVolume(sound.streamId(), sound.volume(), sound.volume());
              }
            }

            viewModel.updateMuteLiveData(muted);
          }
        });

    randomButton.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View v) {

            final Random totalNumberOfSounds = new Random();

            // total number of available sounds can be found in viewmodel in sounds variable
            List<Sound> value = viewModel.sounds().getValue();

            for (int i = 0; i < value.size(); i++) {
              Log.d(TAG, "randomClick for loop called: " + i);
              Sound sound = value.get(totalNumberOfSounds.nextInt(value.size()));
              if (!sound.isPro()) {
                playStopSound(sound.soundPoolId(), sound.isPlaying(), sound.streamId());
              }
            }
          }
        });

    playStopButton.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View v) {
            if (viewModel.playingSounds().getValue() != null
                && viewModel.playingSounds().getValue().size() != 0) {
              // stop all sounds and show play button again
              stopAllSounds();
            }
          }
        });

    saveFavorites.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View v) {

            Boolean atLeastOneIsPlaying = viewModel.isAtLeastOneSoundPlaying().getValue();

            if (atLeastOneIsPlaying != null && atLeastOneIsPlaying) {
              activityCallback.showAddToFavoritesDialog(
                  getSoundParameters(viewModel.playingSounds().getValue()));
            } else {
              activityCallback.showToast("You must play at least one sound");
            }
          }
        });

    setTimer.setOnClickListener(
        new View.OnClickListener() {
          @Override
          public void onClick(View v) {

            Boolean atLeastOneIsPlaying = viewModel.isAtLeastOneSoundPlaying().getValue();

            if (atLeastOneIsPlaying != null && atLeastOneIsPlaying) {

              Boolean timerIsRunning = timerEnabled.getValue();

              if (timerIsRunning != null && timerIsRunning) {
                timerEnabled.setValue(false);
              } else {
                // show TimerDialog fragment created with getFilePath template
                activityCallback.showTimerDialog();
              }

            } else {
              activityCallback.showToast("You must play at least one sound");
            }
          }
        });
  }

  private HashMap<Integer, Integer> getSoundParameters(List<Sound> sounds) {

    HashMap<Integer, Integer> hashMap = new HashMap<>();

    for (Sound sound : sounds) {
      hashMap.put(sound.soundPoolId(), sound.streamId());
    }

    return hashMap;
  }

  // endregion

  private void loadToSoundPool(List<Sound> sounds) {

    Log.d(TAG, "loadToSoundPool: called");

    ArrayList<Sound> sounds1 = new ArrayList<>();

    for (Sound sound : sounds) {

      if (sound.soundPoolId() == 0) {
        // add to arraylist with soundId from soundpool
        int soundId = soundPool.load(sound.getFilePath(), 1);
        sounds1.add(Sound.withSoundPoolId(sound, soundId));
      } else {
        sounds1.add(sound);
      }
    }

    Log.d(TAG, "loadToSoundPool: sounds: " + sounds1);

    viewModel.addToSounds(sounds1);
  }

  private void showTimerText() {
    timerText.setVisibility(View.VISIBLE);
  }

  private void hideTimerText() {
    Log.d(TAG, "hideTimerText: called");
    timerText.setVisibility(View.INVISIBLE);
  }

  private void setTimerText(String text) {
    timerText.setText(text);
  }

  public void startCountDownTimer(Integer minutes) {

    Log.d(TAG, "startCountDownTimer: minutes: " + minutes);

    if (minutes == 999) {
      new TimePickerDialog(getContext(), timerPickerListener, 0, 0, true).show();
    } else {
      toggleCountdownTimer(minutes);
    }
  }

  private void toggleCountdownTimer(int minutes) {

    if (timerEnabled.getValue() != null && timerEnabled.getValue() && countDownTimer != null) {
      timerEnabled.setValue(false);
    } else {
      countDownTimer =
          new CountDownTimer(TimeUnit.MINUTES.toMillis(minutes), 1000) {

            public void onTick(long millisUntilFinished) {

              setTimerText(
                  String.format(
                      "Sound%s will stop in " + TimerDialog.getCountTimeByLong(millisUntilFinished),
                      viewModel.playingSounds().getValue().size() > 1 ? "s" : ""));
            }

            public void onFinish() {
              setTimer.setImageDrawable(getResources().getDrawable(R.drawable.ic_timer_on));
              timerEnabled.setValue(false);
              stopAllSounds();
              hideTimerText();
            }
          }.start();

      timerEnabled.setValue(true);
    }
  }

  public void triggerCombo(SavedCombo savedCombo) {

    boolean areSoundsStillPlaying = stopAllSounds();

    for (Map.Entry<Integer, Integer> entry : savedCombo.getSoundPoolParameters().entrySet()) {
      playStopSound(entry.getKey(), areSoundsStillPlaying, entry.getValue());
    }
  }

  public void fetchSounds() {
    viewModel.fetchSounds();
  }

  @Override
  public void onStart() {
    super.onStart();
    activityCallback.soundGridFragmentStarted();
  }

  public boolean soundsAlreadyFetched() {
    return viewModel.getSoundsAlreadyFetched();
  }
}

File: app/src/main/java/com/cazimir/relaxoo/ui/sound_grid/SoundGridFragment.kt
package com.cazimir.relaxoo.ui.sound_grid

import android.app.TimePickerDialog
import android.app.TimePickerDialog.OnTimeSetListener
import android.content.Context
import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TimePicker
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.SavedStateViewModelFactory
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelStoreOwner
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.cazimir.relaxoo.MainActivity
import com.cazimir.relaxoo.R
import com.cazimir.relaxoo.adapter.GridRecyclerViewAdapter
import com.cazimir.relaxoo.analytics.AnalyticsEvents.Companion.muteClicked
import com.cazimir.relaxoo.analytics.AnalyticsEvents.Companion.randomClicked
import com.cazimir.relaxoo.analytics.AnalyticsEvents.Companion.removeFromDashboard
import com.cazimir.relaxoo.analytics.AnalyticsEvents.Companion.soundClicked
import com.cazimir.relaxoo.analytics.AnalyticsEvents.Companion.timerClicked
import com.cazimir.relaxoo.dialog.custom.BottomCustomDeleteFragment
import com.cazimir.relaxoo.dialog.custom.CustomBottomCallback
import com.cazimir.relaxoo.eventbus.*
import com.cazimir.relaxoo.model.ListOfSavedCustom
import com.cazimir.relaxoo.model.SavedCombo
import com.cazimir.relaxoo.model.Sound
import com.cazimir.relaxoo.model.TimerData
import com.cazimir.relaxoo.service.SoundService
import com.cazimir.relaxoo.service.commands.*
import com.cazimir.relaxoo.shared.SharedViewModel
import com.cazimir.utilitieslibrary.SharedPreferencesUtil.loadFromSharedPreferences
import com.cazimir.utilitieslibrary.SharedPreferencesUtil.saveToSharedPreferences
import com.cazimir.utilitieslibrary.observeOnceOnListEmptyWithOwner
import com.cazimir.utilitieslibrary.observeOnceOnListNotEmptyWithOwner
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.analytics.FirebaseAnalytics
import kotlinx.android.synthetic.main.sound_list_fragment.*
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.*

class SoundGridFragment : Fragment() {

    private lateinit var sharedViewModel: SharedViewModel
    private var timerRunning: Boolean = false
    private var soundsAdapter: GridRecyclerViewAdapter? = null
    private lateinit var viewModel: SoundGridViewModel
    private lateinit var currentlyPlayingSounds: List<Sound>
    private lateinit var activityCallback: OnActivityCallback
    private val timerPickerListener =
            OnTimeSetListener { view: TimePicker?, hours: Int, minutes: Int ->
                if ((hours * 60) + minutes != 0) {
                    toggleCountdownTimer((hours * 60) + minutes)
                } else {
                    activityCallback.showMessageToUser(getString(R.string.custom_timer_zero), Snackbar.LENGTH_SHORT)
                }
            }

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? {
        val view =
                inflater.inflate(R.layout.sound_list_fragment, container, false)
        return view
    }

    private fun playStopSound(sound: Sound) {
        if (sound.playing) {
            sendCommandToService(
                    SoundService.getCommand(
                            context,
                            StopCommand(sound)
                    )
            )
        } else {
            if (sound.loaded) {
                sendCommandToService(
                        SoundService.getCommand(
                                context,
                                PlayCommand(sound)
                        )
                )
            } else {
                activityCallback.showMessageToUser(getString(R.string.playing_except_loading), Snackbar.LENGTH_SHORT)
            }
        }
    }

    override fun onDetach() {
        super.onDetach()
        EventBus.getDefault().unregister(this)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnActivityCallback) {
            activityCallback = context
        }
    }

    private fun stopAllSounds() {
        sendCommandToService(
                SoundService.getCommand(
                        context,
                        StopAllSoundsCommand()
                )
        )
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)


        sharedViewModel = ViewModelProvider(activity as ViewModelStoreOwner).get(SharedViewModel::class.java)

        if (!EventBus.getDefault().isRegistered(this)) {
            EventBus.getDefault().register(this)
        }
        // used to pass the savedstate to ViewModel when application is beeing destroyed by the Android OS
        val factory = SavedStateViewModelFactory(activity!!.application, activity as FragmentActivity)

        viewModel = ViewModelProvider(this, factory).get(SoundGridViewModel::class.java)

        setListenersForButtons()

        sounds_recycler_view.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                val layoutManager = sounds_recycler_view.layoutManager as GridLayoutManager
                sharedViewModel.currentScrollPosition = layoutManager.findFirstVisibleItemPosition()
            }
        })

        // region Observers

        viewModel.initialFetchFinished.observeOnceOnListNotEmptyWithOwner(viewLifecycleOwner, Observer { list: List<Sound> ->
            //get these observables to sync with the service after an application restart
            // using this if because of ROTATION
            // what about when relaunching app from notification and resync is needed

            //check is number of sounds on service same as fetched number of sounds
            sendCommandToService(SoundService.getCommand(context, GetNumberOfSoundsCommand(list)))

        })

        // listen for changes to the sound lists live data object to set the adapter for the gridview
        // along with the callback methods (clicked & volume changed)
        viewModel
                .soundsStorage
                .observe( // TODO: 19.12.2019 move in a separate file or inner class
                        viewLifecycleOwner,
                        Observer { sounds: List<Sound> ->
                            val newList = sounds.toList()
                            updateOrSetupAdapter(newList)

                            // if list same size (only changes on properties like playing, loaded etc)
                            if (newList.size == soundsAdapter?.sounds?.size) {
                                val diff = newList.subtract(soundsAdapter?.sounds as List<Sound>)

                                //no difference between the lists
                                if (diff.isNotEmpty()) {
                                    for ((index, value) in diff.withIndex()) {
                                        soundsAdapter?.modifySingleSoundInList(diff.elementAt(index))
                                    }
                                }
                                // if new list size smaller than what is currently in the adapter remove the 'in plus' item from the adapter
                            } else if (newList.size < soundsAdapter?.sounds?.size!!) {

                                val toDelete = soundsAdapter?.sounds?.find {
                                    !newList.contains(it)
                                }

                                soundsAdapter?.removeSingleSoundInList(toDelete!!)

                            } else {
                                newList.forEach {
                                    if (!soundsAdapter?.sounds?.contains(it)!!) {
                                        soundsAdapter?.sounds?.add(it)
                                    }
                                }
                            }
                        }
                )

        // listen to the playing sounds live data object to change the play stop getLogoPath icon on top
        viewModel
                .playingSounds()
                .observe(
                        viewLifecycleOwner, Observer<List<Sound?>> { playingList ->

                    currentlyPlayingSounds = playingList as List<Sound>

                    if (playingList.isEmpty()) {
                        play_button.setImageDrawable(resources.getDrawable(R.drawable.ic_play))
                        play_button.tag = getString(R.string.play_button_tag)
                        set_timer_button.setImageDrawable(resources.getDrawable(R.drawable.ic_timer_off))
                    } else {
                        play_button.setImageDrawable(resources.getDrawable(R.drawable.ic_stop_white))
                        play_button.tag = getString(R.string.stop_button_tag)
                    }
                })

        viewModel.soundsLoadedToSoundPool
                .observe(
                        viewLifecycleOwner,
                        Observer { soundsAdded ->
                            if (viewModel.soundsStorage.value?.size != 0) {
                                if (soundsAdded == 3) {
                                    activityCallback.hideSplashScreen()
                                }
                            }
                        })
        // endregion
    }

    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    fun callbackWithNumberOfCurrentSounds(lists: EventBusSendNumberOfSounds) {

        // fetched size same as sounds on service. just synchronize
        if (lists.fetchedSounds.size == lists.soundOnService.size) {
            sendCommandToService(SoundService.getCommand(context, AllSoundsCommand()))
            sendCommandToService(SoundService.getCommand(context, TimerTextCommand()))
            sendCommandToService(SoundService.getCommand(context, MuteStatusCommand()))
            //normally you would hide the splash when at least 3 sounds have been loaded to the soundpool. in this case the load happens almost instantaneously. set small delay

            activityCallback.hideSplashScreen()

        } else {
            if (shouldLoadToSoundpool()) {
                loadListToSoundPool(lists.fetchedSounds)
                // because you already did it once per app run
                viewModel.shouldLoadToSoundPool = false
            }

            sendCommandToService(SoundService.getCommand(context, AllSoundsCommand()))
            sendCommandToService(SoundService.getCommand(context, TimerTextCommand()))
            sendCommandToService(SoundService.getCommand(context, MuteStatusCommand()))
        }
    }

    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    fun recordingNameUpdated(eventBusUpdateRecordingName: EventBusUpdateRecordingName) {

        val recordingId = eventBusUpdateRecordingName.id
        val recordingNewName = eventBusUpdateRecordingName.newName

        val findRecording = viewModel.soundsStorage.value?.find { it.id == recordingId }

        findRecording?.let {
            removeCustomSoundFromDashboardIfThere(recordingId)

            val sound = Sound(custom = true,
                    filePath = "/storage/emulated/0/Relaxoo/own_sounds/$recordingNewName.wav",
                    logoPath = "/storage/emulated/0/Relaxoo/logos/thunder.png",
                    name = recordingNewName,
                    id = "$recordingNewName.wav")

            activityCallback.pinToDashBoardActionCalled(sound)
        }
    }

    private fun updateOrSetupAdapter(allSounds: List<Sound>) {
        if (soundsAdapter == null) {
            soundsAdapter = GridRecyclerViewAdapter(context!!,
                    allSounds as ArrayList<Sound>,
                    object : OnSoundClickListener {
                        override fun clicked(sound: Sound) {
                            FirebaseAnalytics.getInstance(context!!).logEvent(soundClicked(sound).first, soundClicked(sound).second)
                            if (!sound.loaded) {
                                activityCallback.showMessageToUser(getString(R.string.sound_loading), Snackbar.LENGTH_SHORT)
                            } else if (sound.pro && !sound.playing) {
                                viewModel.setClickedProSound(sound)
                                activityCallback.showBottomDialogForPro()
                            } else {
                                playStopSound(sound)
                            }
                        }

                        override fun volumeChange(
                                sound: Sound,
                                volume: Int
                        ) { // TODO: 18.12.2019 refactor this float to string to double transformation
                            val volumeToSet: Float =
                                    (volume.toDouble() / 100).toString().toFloat()

                            val volumeCommand = SoundService.getCommand(
                                    context, VolumeCommand(sound.id,
                                    sound.streamId,
                                    volumeToSet,
                                    volumeToSet
                            )
                            )

                            sendCommandToService(volumeCommand)
                        }

                        override fun volumeChangeStopped(
                                sound: Sound,
                                progress: Int
                        ) {
                            viewModel.updateVolume(
                                    sound, (progress.toDouble() / 100).toString().toFloat()
                            )
                        }

                        override fun moreOptionsClicked(sound: Sound) {
                            BottomCustomDeleteFragment(sound, object : CustomBottomCallback {
                                override fun deletedClicked(sound: Sound) {
                                    FirebaseAnalytics.getInstance(context!!).logEvent(removeFromDashboard().first, removeFromDashboard().second)
                                    removeRecordingFromSoundPool(sound)
                                }
                            }).show(parentFragmentManager, "deleteCustom")
                        }
                    })

            val numberOfColumns = if (resources.configuration.orientation == Configuration.ORIENTATION_PORTRAIT) 3 else 4

            val gridLayoutManager = GridLayoutManager(context, numberOfColumns)
            sounds_recycler_view.layoutManager = gridLayoutManager
            sounds_recycler_view.adapter = soundsAdapter

            sounds_recycler_view.smoothScrollToPosition(sharedViewModel.currentScrollPosition)

        }
    }

    // region Listeners for buttons on top
    private fun setListenersForButtons() {
        mute_button.setOnClickListener {
            FirebaseAnalytics.getInstance(context!!).logEvent(muteClicked().first, muteClicked().second)
            //if no sound is playing do not allow user to mute
            if (currentlyPlayingSounds.isNotEmpty()) {
                sendCommandToService(
                        SoundService.getCommand(
                                context,
                                ToggleMuteCommand()
                        )
                )
            } else {
                activityCallback.showMessageToUser(getString(R.string.play_one_sound), Snackbar.LENGTH_SHORT)
            }
        }
        random_button.setOnClickListener {
            FirebaseAnalytics.getInstance(context!!).logEvent(randomClicked().first, randomClicked().second)
            stopAllSounds()

            activityCallback.showMessageToUser(getString(R.string.playing_random), Snackbar.LENGTH_SHORT)

            /*observe playing sounds once to make sure that an empty list is delivered and the observe once and remove observer*/
            viewModel.playingSounds().observeOnceOnListEmptyWithOwner(viewLifecycleOwner, Observer {
                // total number of available sounds can be found in viewmodel in sounds variable
                val listAllSounds: List<Sound> = (viewModel.soundsStorage.value)!!

                val processed = mutableListOf<Sound>()

                while (processed.size < calculateLimit(listAllSounds)) {
                    val randomSound = listAllSounds.random()
                    if (randomSound.loaded && !randomSound.pro && !randomSound.playing && !processed.contains(randomSound)
                    ) {
                        playStopSound(randomSound)
                        processed.add(randomSound)
                    }
                }
            })
        }
        play_button.setOnClickListener {
            if (currentlyPlayingSounds.isNotEmpty()) {
                stopAllSounds()
            }
        }
        save_fav_button.setOnClickListener {
            val atLeastOneIsPlaying = viewModel.soundsStorage.value?.find { sound -> sound.playing }
            if (atLeastOneIsPlaying != null) {
                activityCallback.showAddToFavoritesDialog(currentlyPlayingSounds)
            } else {
                activityCallback.showMessageToUser(getString(R.string.play_one_sound), Snackbar.LENGTH_SHORT)
            }
        }

        set_timer_button.setOnClickListener {
            FirebaseAnalytics.getInstance(context!!).logEvent(timerClicked().first, timerClicked().second)
            val atLeastOneIsPlaying = currentlyPlayingSounds.isNotEmpty()
            if (atLeastOneIsPlaying) {
                if (timerRunning) {
                    stopAllSounds()
                } else {
                    activityCallback.showTimerDialog()
                }
            } else {
                activityCallback.showMessageToUser(getString(R.string.play_one_sound), Snackbar.LENGTH_SHORT)
            }
        }
    }

    private fun calculateLimit(listSize: List<Sound>): Int {

        val notPro = listSize.filter { !it.pro }
        if (notPro.size < MAX_RANDOM_SOUNDS) {
            return notPro.size
        }

        return MAX_RANDOM_SOUNDS
    }

    // endregion
    private fun loadListToSoundPool(sounds: List<Sound>) {
        sendCommandToService(
                SoundService.getCommand(
                        context,
                        LoadSoundsCommand(sounds)
                )
        )
    }

    private fun showTimerText() {
        timerText.visibility = View.VISIBLE
    }

    private fun hideTimerText() {
        timerText.visibility = View.GONE
    }

    private fun setTimerText(text: String) {
        timerText.text = text
    }

    fun startCountDownTimer(minutes: Int) {
        if (minutes == 999) {
            TimePickerDialog(context, timerPickerListener, 0, 0, true).show()
        } else {
            toggleCountdownTimer(minutes)
        }
    }

    private fun toggleCountdownTimer(minutes: Int) {
        sendCommandToService(SoundService.getCommand(context, ToggleCountDownTimerCommand(minutes)))
    }

    fun triggerCombo(savedCombo: SavedCombo, boughtPro: Boolean?) {
        sendCommandToService(SoundService.getCommand(context, TriggerComboCommand(savedCombo.sounds, boughtPro ?: false)))
    }

    fun fetchSoundsOnline() {
        viewModel.fetchSounds()
    }

    fun fetchSoundsOffline() {
        viewModel.fetchSoundsOffline()
    }

    override fun onResume() {
        super.onResume()
        activityCallback.soundGridFragmentStarted()
    }

    fun shouldLoadToSoundpool(): Boolean {
        return viewModel.shouldLoadToSoundPool
    }

    private fun removeRecordingFromSoundPool(sound: Sound) {
        sendCommandToService(
                SoundService.getCommand(
                        context,
                        UnloadSoundCommand(sound)
                )
        )
    }

    private fun scrollToBottom() {
        sounds_recycler_view!!.smoothScrollToPosition(soundsAdapter!!.itemCount)
    }

    fun rewardUserByPlayingProSound() {
        // if not null
        viewModel.currentlyClickedProSound?.let { sound ->
            playStopSound(sound)
        }
    }

    fun sendCommandToService(intent: Intent) {
        context?.startService(intent)
    }

    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    fun updateViewModelWithLoad(eventBusLoad: EventBusLoad) {
        viewModel.loadedToSoundPool()
    }

    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    fun updateViewModelWithUnload(eventBusUnload: EventBusUnload) {
        val newList = mutableListOf<Sound>()
        val pinnedRecordings = loadFromSharedPreferences<ListOfSavedCustom>(MainActivity.PINNED_RECORDINGS)
        val list = pinnedRecordings?.savedCustomList ?: mutableListOf()

        list.filterTo(newList, {
            it.id != eventBusUnload.sound.id
        })

        val newObject = ListOfSavedCustom(newList)
        saveToSharedPreferences(newObject, MainActivity.PINNED_RECORDINGS)
    }

    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    fun updateTimerLiveDataInViewModel(eventBusTimerStarted: EventBusTimer) {
        // no need to put this livedata in viewmodel because it is being observed from a foreground service which never dies unless the user shuts it down.
        eventBusTimerStarted._timer
                .observe(
                        viewLifecycleOwner,
                        Observer { timerData: TimerData ->

                            if (timerData.timerRunning) {
                                showTimerText()
                                setTimerText(timerData.timerText)
                                set_timer_button.setImageDrawable(resources.getDrawable(R.drawable.ic_timer_on))
                                timerRunning = true
                            } else {
                                hideTimerText()
                                setTimerText(timerData.timerText)
                                set_timer_button.setImageDrawable(resources.getDrawable(R.drawable.ic_timer_off))
                                timerRunning = false
                            }
                        }
                )
    }

    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    fun singleSoundLoadedToSoundpoolCallback(eventBusLoad: EventBusLoadSingle) {
        scrollToBottom()
    }

    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    fun muteStatusReceived(eventBusLoad: EventBusMuteStatus) {
        eventBusLoad.muteStatus.observe(viewLifecycleOwner, Observer {
            if (it) {
                mute_button.setImageDrawable(resources.getDrawable(R.drawable.ic_mute_on))
                // used for espresso
                mute_button.tag = getString(R.string.mute_on_tag)
            } else {
                mute_button.setImageDrawable(resources.getDrawable(R.drawable.ic_mute_off_white))
                // used for espresso
                mute_button.tag = getString(R.string.mute_off_tag)
            }
        })
    }


    fun removeCustomSoundFromDashboardIfThere(recordingId: String) {
        val filtered = mutableListOf<Sound>()
        viewModel.soundsStorage.value?.filterTo(filtered, predicate = {
            it.id == recordingId
        })

        // i imagine there will be only one
        if (filtered.size != 0) {
            //found recording in the sounds store
            removeRecordingFromSoundPool(filtered.first())
        }
    }

    fun areSoundsStillLoading(): Boolean {

        var areLoading = false

        if (viewModel.soundsStorage.value?.find { sound -> !sound.loaded } != null) {
            areLoading = true
        }

        return areLoading
    }

    companion object {
        private val TAG = "SoundGridFragment"
        private val MAX_RANDOM_SOUNDS = 3

        fun newInstance(): SoundGridFragment {
            return SoundGridFragment()
        }
    }
}
